### ISSUE TITLE
Proposal: Automatically Drop Old Remote Write Samples Before Sending

### ISSUE THREAD

**Issue by vsliouniaev** [prometheus/prometheus#7912]
## Proposal
When a remote-write endpoint is unavailable for a long time and then comes back, the data prometheus has may be older than the endpoint will allow to be remote-written, due to various internal implementation reasons.

If this happens, prometheus will try to keep sending data that will never be accepted until it the endpoint returns, at which point it will be rejected. Due to the volume of samples as well as the inherently shorter code-path common for sample rejection scenarios this may result in saturation of both network bandwidth on the prometheus side and bandwidth as well as memory on the receive side.

A parameter in the remote-write configuration that would allow discarding of samples older than some pre-defined limit, rather than attempting to send them would help in this scenario.

**Comment by roidelapluie** [comment]
Thanks.

cc @csmarchbanks @cstyan What do you think about this request?

**Comment by csmarchbanks** [comment]
Thanks for the proposal!

We just had a discussion around this topic in a Cortex community call as well. I would like a solution such as what is proposed here. I have had to manually restart Prometheus instances before to discard the backlog to avoid rate limits or other issues which is not fun. https://github.com/prometheus/prometheus/issues/7078 also discusses some issues around Cortex never being able to catch up after cutting a block that this would solve.

I like the idea of using time, but I worry a bit about what would happen to samples that are supposed to be offset? We only store the timestamp exposed by a sample, not the timestamp when we collected the sample which could cause some unintuitive behavior. Maybe we don't need to worry about that case though? E.g. someone who is running cloudwatch exporter with a 15m offset, and configures remote write to only send samples newer than 10m would never receive their cloudwatch exporter metrics. I imagine this is rather advanced usage, so maybe this isn't too much of a concern and could be documented?

Another option could be a certain amount of samples/data behind, but that is harder to reason about in the majority of cases.

There were also some ideas that came up in the Cortex call around using headers to control behavior (would be helpful for rate limiting as well). We could have a set of Prometheus headers and one of them represents the earliest timestamp that the destination can accept?

TLDR; Generally I am for something like this, just wanted to provide a couple options/thoughts and context around other discussions.

**Comment by brian-brazil** [comment]
Is this a problem with Prometheus, or a problem with a particular remote storage? Prometheus doesn't seem to be doing anything wrong here so I'd not immediately jump to working around issues in remote storage by adding config and code complexity in Prometheus.

> We could have a set of Prometheus headers and one of them represents the earliest timestamp that the destination can accept?

That sounds like negotiation, and remote-write is a stateless protocol.

**Comment by vsliouniaev** [comment]
The behaviour that we are looking for could be solved by adding a small proxy which would perform the operation, but I believe the case of samples not being accepted beyond a certain time is a universal one with remote storage systems, so it seemed like a reasonable configuration option. The scenario which prompted this proposal involved hundreds of prometheuses running in very resource-constrained environments that would be helped by not having to decompress and re-compress requests

**Comment by brian-brazil** [comment]
It's Prometheus that defines the protocol, so if the remote storage breaks on valid input I'd suggest that in the first instance it is the remote storage which needs to change their behaviour.

**Comment by csmarchbanks** [comment]
> Is this a problem with Prometheus, or a problem with a particular remote storage?

The problem with Prometheus is that it can be hard to recover remote write after a significant period of downtime in the remote system, sometimes resulting in manual intervention to intentionally drop data. When trying to catch up it is very likely to come across rate limits, issues ingesting old data, timeouts as every Prometheus instance tries to catch up simultaneously, etc... 

At a minimum, I think we should allow users to intentionally choose between having gaps in their data (especially if their remote system won't accept the data anyway), and having more recent data available in the remote system.

> > We could have a set of Prometheus headers and one of them represents the earliest timestamp that the destination can accept?
>
> That sounds like negotiation, and remote-write is a stateless protocol.

The content would still be the same and able to be processed in isolation. I think having hints to help control sharding, backoffs, when to drop data, etc... might be nice, but I am also not sure if that is a pattern we should support. I think someone from the Cortex call was going to write up more thoughts around this idea.

**Comment by brian-brazil** [comment]
> The problem with Prometheus is that it can be hard to recover remote write after a significant period of downtime in the remote system, 

I'm not seeing why you think this is a problem with Prometheus, all of the issues you list appear to be entirely on the remote storage end.

> At a minimum, I think we should allow users to intentionally choose between having gaps in their data (especially if their remote system won't accept the data anyway), and having more recent data available in the remote system.

That doesn't sound completely unreasonable. Might the checkpoint stuff let us cheat there?

**Comment by csmarchbanks** [comment]
> I'm not seeing why you think this is a problem with Prometheus, all of the issues you list appear to be entirely on the remote storage end.

Some yes, some no. Rate limiting could be considered a Prometheus issue that we don't handle well. Timeouts can come from either side, especially in resource constrained environments.

> That doesn't sound completely unreasonable. Might the checkpoint stuff let us cheat there?

Which checkpoint stuff are you referring to? The remote write WAL location checkpoints, or the Chunk/WAL snapshots or something else?

**Comment by brian-brazil** [comment]
> Rate limiting could be considered a Prometheus issue that we don't handle well. 

We retry, which seems reasonable to me - do we not have backoff there yet?

> Timeouts can come from either side, especially in resource constrained environments.

If you are getting a timeout on the Prometheus end that implies that the Prometheus is underprovisioned for CPU by a fair bit, so not something I'd worry about much as that Prometheus will already be unhealthy in more typical circumstances.

> Which checkpoint stuff are you referring to? The remote write WAL location checkpoints, 

Yes, the remote write WAL location checkpoints.

**Comment by cstyan** [comment]
It's an issue for the remote storage system via reuse of Prometheus TSDB code, which blocks ingestion of samples older than max timestamp for the head block.

I agree with Chris that we should allow users to choose whether they want to attempt to recover all possible samples in the event of a long outage, or skip to newer samples and have a gap in their remote data.

**Comment by roidelapluie** [comment]
We have got a similar request in #7938.

**Comment by Harkishen-Singh** [comment]
Are we planning to go ahead with this issue? Personally, I liked the concept of headers from prometheus remote write. I also think how would it be if the remote storage can simply respond that the batch sent as some samples that are beyond a particular timestamp, required for acceptable? Then, remote write can simply drop those samples beyond that timestamp?

**Comment by javasparkle** [comment]
We ran into this issue multiple times and the amount of time it takes for Prometheus to clear the outstanding WAL before pushing new samples to remote storage takes it off the table to monitor active incidents.

A header in the remote-write configuration that would allow discarding of samples older than a limit or time will be helpful. Also, another option is to consider prioritizing sending new samples while pushing the old samples in back channel.

**Comment by Harkishen-Singh** [comment]
Then I think, this issue should be considered as a priority.

**Comment by vsliouniaev** [comment]
> Are we planning to go ahead with this issue? Personally, I liked the concept of headers from prometheus remote write. I also think how would it be if the remote storage can simply respond that the batch sent as some samples that are beyond a particular timestamp, required for acceptable? Then, remote write can simply drop those samples beyond that timestamp?

I don't think it's necessary to do this, and as Brian said:

> That sounds like negotiation, and remote-write is a stateless protocol.

I think a configuration on the prometheus-side to define max age of a sample relative to the current time would be the way to address it.

**Comment by csmarchbanks** [comment]
I think configuration on prometheus-side to define a max age of of a sample would provide the most immediate value and least complexity. I would certainly still like to see this, and Callum was also supportive in his last comment, so a PR would be welcome from my point of view.

We cannot prioritize sending newer samples as a guarantee of remote write is that samples are ordered.

**Comment by Harkishen-Singh** [comment]
In that case i will take this up, as the existing PR works are done.

**Comment by cstyan** [comment]
> Also, another option is to consider prioritizing sending new samples while pushing the old samples in back channel.

This is likely not possible, to my knowledge, as many remote storage systems are reusing Prometheus TSDB code which will explicitly reject samples for a series that are older than the last ingested sample.

**Comment by Anhui-tqhuang** [comment]
we have the same problem, have to restart the prometheus manually if remote server goes done for a long time and reject old metrics sent by prometheus, this will also lead a problem: 

new metrics will be blocked due to old metrics is being handled forever.

**Comment by Harkishen-Singh** [comment]
Do we want a TTL in terms of retries, like retry maximum for 5 times and make this configurable through the `queue_configs`?

**Comment by vsliouniaev** [comment]
> Do we want a TTL in terms of retries, like retry maximum for 5 times (default) and make this configurable through the queue_config as max_retries?

I'm not sure what this would achieve. The issue is with sending a sample that is known to be too old for a remote-write endpoint. A well-behaved endpoint receiving remote-writes is able to either reject these with a 4xx status code or simply drop them. In both cases prometheus's remote-write behavior would not attempt a retry.

**Comment by Harkishen-Singh** [comment]
For those following this discussion, here is the [design doc](https://docs.google.com/document/d/1pC5GjMkSe4YHFqVMp0ZX9C1Iih8QLOnnBG-HguKzwrM/edit?usp=sharing) on solving this issue. Please feel free to comment on your suggestions/views/feedbacks.

**Comment by bboreham** [comment]
Reviewing this discussion, I think a couple of points were missed.
First, Prometheus doesn't keep samples indefinitely - the WAL is truncated each time a new block is cut (by default every 2 hours). So "old" is bounded by that time.

Second, the proposal in the design doc is that the remote-write receiver send back what range of timestamps is acceptable, so the sender can (if it chooses) quickly discard those in the queue outside the range. This seems like a good idea to me.

**Comment by davidyuanfs** [comment]
Any update on this proposal? Our team has run into the same issue many times and recovering from a long outages is extremely hard since burning these old records is time consuming and impact real-time data. It always require manually restart, but operational effort is expensive and risky when maintaining in a large scale system. 

I may agree with @vsliouniaev points that adding retries will not achieve the goal we want, since TSDB returning 400 error for old records which remote write doesn't retry.

> Second, the proposal in the design doc is that the remote-write receiver send back what range of timestamps is acceptable, so the sender can (if it chooses) quickly discard those in the queue outside the range. This seems like a good idea to me.

This is also a good idea, but I am thinking what would be the challenge if we just add time range configuration in the remote write, and sender discard the samples which is older than that time range?  Please let us know if there is anything we can help or contribute on this feature, this would be an important feature to improve our system stability. Thank you!

**Comment by jmcarp** [comment]
For what it's worth, the opentelemetry [remote write exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter) includes a `[max_elapsed_time](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md)` setting that I think would be useful here: if a batch has failed to send for $max_elapsed_time seconds, discard it instead retrying until the next WAL compaction.

I would also be happy to write a patch if it would help, if there's an approach that the maintainers want to pursue.

**Comment by cstyan** [comment]
The issue here isn't a batch being retried endlessly. If samples are too old for the receiver to accept they should return a non-recoverable error. The issue is the backlog of samples that need to be read from the WAL and buffered in remote write to be sent. It takes a long time to get through that backlog with the added roundtrip per batch. If remote write simply rejects samples that are too old, or accepts them but drops them instead of buffering them into a batch, we can recover faster.

**Comment by davidyuanfs** [comment]
@callumj exactly! Dropping can be an optional to enable and nice to have metrics about dropping records for system to monitor and alert.

**Comment by davidyuanfs** [comment]
@callumj how about we just add a config in `queue_config` e.g: `[ max_samples_age: <duration> | default = 0s ]`, 0s seconds means no discarding. And for each sample consuming into shard we check and decide based on the sample age before buffering? In the meanwhile we count and emit discard records metric. Would this be a simple and reasonable solution?

If you guys think this is the approach to pursue, we can take it and write a simple proposal doc and send to mailing list to review. And we would like to write the patch.

**Comment by callumj** [comment]
@davidyuanfs I think you are tagging the wrong person?

**Comment by csmarchbanks** [comment]
I am still happy with a solution as described with a max_samples_age. I think there is some disagreement around configuring `max_samples_age` vs a number of retries, but I don't see why we can't have both if someone needs retry based dropping of data.

Note that we should check sample age at a minimum when pulling out of the buffer (in case a sample has sat in the buffer for a long time), and should probably time out a request when the samples in that request become too old based on the configuration. Checking before putting the sample into the buffer would be a nice optimization but not strictly required IMO.

**Comment by davidyuanfs** [comment]
@csmarchbanks Good point about the "check sample age at a minimum", like we should set a 1min at a minimum to avoid overhead time in the buffer.
And also sounds good to time out the batch request if samples in the request have became too old before sending. 

And yes, for the retries, from our point of view, I think dropping data should not couple with retries policy. Dropping old data should be applied to all type of remote write: data haven't sent, sent but return 5xx, 4xx/429(some of the cases need to retry). So I feel we can decouple these two, and in the future if any requirement need to add retry policy, the dropping old data should not block others to add it in my opinion, what do you think?

**Comment by davidyuanfs** [comment]
@callumj sorry for a wrong mention.

**Comment by cstyan** [comment]
> @csmarchbanks Good point about the "check sample age at a minimum", like we can set a 1min at a minimum to avoid overhead time in the buffer. And also sounds good to time out the batch request if samples in the request have became too old before sending.
> 
> And yes, for the retries, from our point of view, I think dropping data should not couple with retries policy. Dropping old data should be applied to all type of remote write data: data haven't sent, sent but return 5xx, 4xx/429(some of the cases need to retry). So I feel we can decouple these two. If any requirement need to add retry policy In the future, the dropping old data should not block others to add it in my opinion. And it is not a breaking change. what do you think?

Chris was saying that we should check if the sample is too old when moving it from the buffer to the next batch of samples to send. That would guarantee we never attempt to send samples that are too old. As an optimization we could _also_ check the sample age before putting the samples we read from the WAL into the buffer in the first place.

**Comment by till** [comment]
Is this proposal going anywhere? Or has anyone thought of workarounds?

**Comment by cstyan** [comment]
There's no active work here at the moment.

**Comment by mvadu** [comment]
@csmarchbanks  in #10785 you indicated core members are still interested to get this feature, but no action for quite some time. We recently got hit with this issue, and having to restart thousands of remote agents (Prometheus code running as part of Grafana agent) where agent is send only mode (i.e. we can't send a command to restart it) was not fun! Do you need volunteers to update #8549 to resolve merge conflicts? or the approach in that PR itself (or the linked design document) itself is not acceptable?

**Comment by csmarchbanks** [comment]
I am still happy to accept a feature that drops old samples before sending. I was on holiday for the last dev summit and did not discuss it there, but if @cstyan doesn't have any opposition to this feature I would say we could move forward.

Updating #8549 would be one option, feel free to at least use it as inspiration. I think the logic around treating exemplars separately is not needed (we should also not send exemplars if they are too old) which would hopefully simplify the PR.

**Comment by milonjames** [comment]
Looking forward to this feature.

**Comment by girish97115** [comment]
What is the update on this feature ?

**Comment by bboreham** [comment]
@girish97115 nobody has coded it.  If there was a PR it would show up as a link here; there is no value in adding a comment which sends email to a dozen people.
You can signify your interest by clicking the 👍 icon on the description, which does not send any email.
